internal/entitlement/aws.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2024 Stefan Prodan.
+// SPDX-License-Identifier: AGPL-3.0
+
+package entitlement
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/marketplacemetering"
+	"github.com/golang-jwt/jwt/v4"
+)
+
+const (
+	// AmazonMarketplaceProductCode is the AWS Marketplace
+	// product code of ControlPlane Enterprise for Flux CD.
+	AmazonMarketplaceProductCode = "272knt6mdmtctbck10givwm1h"
+	// AmazonMarketplacePublicKeyVersion is the AWS Marketplace
+	// public key version of ControlPlane Enterprise for Flux CD.
+	AmazonMarketplacePublicKeyVersion = 1
+	// AmazonMarketplacePublicKey is the AWS Marketplace
+	// public key of ControlPlane Enterprise for Flux CD.
+	AmazonMarketplacePublicKey = `-----BEGIN PUBLIC KEY-----
+MIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEAnFKsHLJY5om6uta6LfG/
+3tfnLO08ZRHeZrromJSEyG+/zdmVJ7s8thk4JQzNxt8fUzgMvKdPuWl17vYay18P
+pCiypg6EbihzIO3VOQCGp1bOSnHrvUlyoyhDuNnG8213DFbl4+MmVwEkI4F25sUq
+56uwmZyHc77ZsjvvFs0pcJ0VQ+DhG0LSjUMmtukeh2VQ29yuQCiKCML4JOkwIRuP
+cmetwvbgn1ViFWwSrhE2i/cNjBzXdd1kz23rmLM4rx4LctsUSAIP3I5YRy4wLUiG
+q+M3YOAcfxQP3t5cjN7rRfyE/bUz+BvipKEPCoDMKmbbNRyX9WYPeIrRsW4HJWGj
+GO0dKDZJJwhMa5TM5zVwepfLeGakxprL7j+0EGFWvf8M0+qZ9OGgEFNVwDVu2BoH
+d1prsH3fI7CTCztrIBgCwtqBhQ5wxzlrnBrDy4WA+CwLFhW77Tghw1E62Vcpj/v5
+vLhgbv3IpMBX2ugEWgeB2i0yWYCpheC8lkbgI90SY+GFAgMBAAE=
+-----END PUBLIC KEY-----`
+)
+
+// AmazonClient is an entitlement client for the
+// ControlPlane Enterprise for Flux CD AWS Marketplace product.
+// https://aws.amazon.com/marketplace/pp/prodview-ndm54wno7tayg
+type AmazonClient struct {
+	Vendor string
+	mc     *marketplacemetering.Client
+}
+
+// NewAmazonClient creates a new AmazonClient using the default
+// AWS configuration and the current region.
+func NewAmazonClient(vendor string) (*AmazonClient, error) {
```

Comment:
It would be nice to have a test for this. I believe we can use `WithEndpointResolverV2()` for that purpose.

```golang
type endpointResolverV2 struct {
	URL url.URL
}

func (e *endpointResolverV2) ResolveEndpoint(ctx context.Context, params marketplacemetering.EndpointParameters) (smithyendpoints.Endpoint, error) {
	return smithyendpoints.Endpoint{
		URI: e.URL,
	}, nil
}

func TestMarketplaceClient() {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		...
	}))
	defer srv.Close()
	opt := marketplacemetering.WithEndpointResolverV2(&endpointResolverV2{URL: url.URL{Scheme: "http", Host: srv.URL}})
	mc := marketplacemetering.NewFromConfig(cfg, opt)
	...
}
```

Reply:
Mocking the AWS APIs could come later after we release a version with metering and we tests it on EKS.

Reply:
I've tested this PR on EKS with an IRSA for `AWSMarketplaceMeteringRegisterUsage` and all works, the usage is registered and the resulting token is valid 🎉